util/metrics/prometheus.py

```diff
@@ -26,159 +33,66 @@ def __init__(self, app=None):
+            # Practically disable this worker, if there is no pushgateway.
+            if agg_url is None or os.getenv("TEST") != "":
+                time.sleep(ONE_DAY_IN_SECONDS)
+                continue
 
+            time.sleep(PROMETHEUS_PUSH_INTERVAL_SECONDS)
             try:
-                resp = requests.post(self._url + "/call", "\n".join(reqs))
-                if resp.status_code == 500 and self.next_register <= datetime.datetime.now():
-                    resp = requests.post(self._url + "/call", "\n".join(self._metric_collectors))
-                    self.next_register = datetime.datetime.now() + REGISTER_WAIT
-                    logger.debug(
-                        "Register returned %s for %s metrics; setting next to %s",
-                        resp.status_code,
-                        len(self._metric_collectors),
-                        self.next_register,
-                    )
-                elif resp.status_code != 200:
-                    logger.debug(
-                        "Failed sending to prometheus: %s: %s: %s",
-                        resp.status_code,
-                        resp.text,
-                        ", ".join(reqs),
-                    )
-                else:
-                    logger.debug("Sent %d prometheus metrics", len(reqs))
-            except:
-                logger.exception("Failed to write to prometheus aggregator: %s", reqs)
-
-
-class _Collector(object):
-    """ Collector for a Prometheus metric. """
-
-    def __init__(
-        self,
-        enqueue_method,
-        collector_type,
-        collector_name,
-        collector_help,
-        namespace="",
-        subsystem="",
-        **kwargs
-    ):
-        self._enqueue_method = enqueue_method
-        self._base_args = {
-            "Name": collector_name,
-            "Namespace": namespace,
-            "Subsystem": subsystem,
-            "Type": collector_type,
-        }
-
-        registration_params = dict(kwargs)
-        registration_params.update(self._base_args)
-        registration_params["Help"] = collector_help
-
-        self._enqueue_method("register", registration_params)
-
-    def __getattr__(self, method):
-        def f(value=0, labelvalues=()):
-            data = dict(self._base_args)
-            data.update(
-                {"Value": value, "LabelValues": [str(i) for i in labelvalues], "Method": method,}
-            )
-
-            self._enqueue_method("put", data)
+                push_to_gateway(
+                    agg_url,
+                    job=self._app.config.get("PROMETHEUS_NAMESPACE", "quay"),
+                    registry=REGISTRY,
+                )
+            except urllib2.URLError:
+                # There are many scenarios when the gateway might not be running.
+                # These could be testing scenarios or simply processes racing to start.
+                # Rather than try to guess all of them, keep it simple and let it fail.
+                pass
```

Comment:
Should it be logged though?

Reply:
I considered this, but the signal to noise on things like running tests is non-ideal.
I'm totally open to suggestions for this. It's honestly the last logical change I've made this PR because the tests run in various modes.

Reply:
Aren't the tests being skipped with the check above?

Reply:
I don't think catches the e2e registry tests and when you run stuff like initdb :|